components/script/dom/worklet.rs

```diff
@@ -0,0 +1,621 @@
+    pub fn new(window: &Window, global_type: WorkletGlobalScopeType) -> Root<Worklet> {
+        debug!("Creating worklet {:?}.", global_type);
+        let worklet = box Worklet::new_inherited(window, global_type);
+        let global = window.upcast::<GlobalScope>();
+        unsafe {
+            WorkletBinding::Wrap(global.get_cx(), global, worklet)
+        }
+    }
+
+    pub fn worklet_id(&self) -> WorkletId {
+        self.worklet_id
+    }
+
+    #[allow(dead_code)]
+    pub fn worklet_global_scope_type(&self) -> WorkletGlobalScopeType {
+        self.global_type
+    }
+}
+
+impl WorkletMethods for Worklet {
+    #[allow(unrooted_must_root)]
+    /// https://drafts.css-houdini.org/worklets/#dom-worklet-addmodule
+    fn AddModule(&self, module_url: USVString, options: &WorkletOptions) -> Rc<Promise> {
+        // Step 1.
+        let promise = Promise::new(self.window.upcast());
+
+        // Step 3.
+        let module_url_record = match self.window.Document().base_url().join(&module_url.0) {
+            Ok(url) => url,
+            Err(err) => {
+                // Step 4.
+                debug!("URL {:?} parse error {:?}.", module_url.0, err);
+                promise.reject_error(self.window.get_cx(), Error::Syntax);
+                return promise;
+            }
+        };
+        debug!("Adding Worklet module {}.", module_url_record);
+
+        // Steps 6-12 in parallel.
+        let pending_tasks_struct = PendingTasksStruct::new();
+        let global = self.window.upcast::<GlobalScope>();
+        let pool = ScriptThread::worklet_thread_pool();
+        pool.fetch_and_invoke_a_worklet_script(global.pipeline_id(),
+                                               self.worklet_id,
+                                               self.global_type,
+                                               global.api_base_url(),
+                                               module_url_record,
+                                               options.credentials.clone(),
+                                               pending_tasks_struct,
+                                               &promise);
+
+        // Step 5.
+        promise
+    }
+}
+
+/// A guid for worklets.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, JSTraceable)]
+pub struct WorkletId(Uuid);
+
+known_heap_size!(0, WorkletId);
+
+impl WorkletId {
+    fn new() -> WorkletId {
+        WorkletId(servo_rand::random())
+    }
+}
+
+/// https://drafts.css-houdini.org/worklets/#pending-tasks-struct
+#[derive(Clone, Debug)]
+struct PendingTasksStruct(Arc<AtomicIsize>);
+
+impl PendingTasksStruct {
+    fn new() -> PendingTasksStruct {
+        PendingTasksStruct(Arc::new(AtomicIsize::new(WORKLET_THREAD_POOL_SIZE as isize)))
+    }
+
+    fn set_counter_to(&self, value: isize) -> isize {
+        self.0.swap(value, Ordering::AcqRel)
+    }
+
+    fn decrement_counter_by(&self, offset: isize) -> isize {
+        self.0.fetch_sub(offset, Ordering::AcqRel)
```

Comment:
Using the weakest ordering that does the job. In this case it doesn't matter a huge amount, this isn't hot code.